src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,171 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::fabric::Network,
+    protocol::{
+        context::ProtocolContext,
+        reveal::reveal,
+        sort::ApplyInvStep::{RevealPermutation, ShuffleInputs, ShufflePermutation},
+        RecordId,
+    },
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use futures::future::try_join_all;
+use permutation::Permutation;
+
+use super::{
+    apply::apply,
+    shuffle::{generate_random_permutation, Shuffle},
+};
+use tokio::try_join;
+
+/// This is APPLYINV(Algorithm 4) described in <https://eprint.iacr.org/2019/695.pdf>.
+/// This is a protocol that applies the inverse of a shared permutation to a shared-vector representation.
+/// Input: Each helpers know their own secret shares of input and permutation
+/// Output: At the end of the protocol, all helpers receive inputs after the permutation is applied
+#[derive(Debug)]
+#[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+pub struct SecureApplyInv<'a, F> {
+    input: &'a mut Vec<Replicated<F>>,
+    permutation: &'a mut Vec<Replicated<F>>,
```

Comment:
qq just to clarify my understanding - does permutation need to be in the same field as input? I assume that is just secret shared indices, so is it possible that we may decide to use $p = 2^{61}$ for element field and $p = 2^{31}$ for permutation? Does it make sense or is it even possible to take two different fields here `F1` and `F2`?

Reply:
Makes sense.. let me change this

Reply:
For now, I will let the protocol take in two fields F1 and F2 for input and permutation respectively. SecretSharing change can be handled on top of it?

Reply:
that's fine. If you want to do that now - OK. If you want to wait until later, I'm fine with that too.

Reply:
ACTUALLY, I cant do this. With the new changes in SecurityValidatorAccumulator, it expects a single type of F whereas for validating shuffles we would need two F's (one for input and one for permutation). Lets discuss this offline.
For this PR, leaving it as F.

Reply:
Ok, let’s leave it for now - and discuss this later.